DatacenterIntegration/Portal/PortalUserDataGdprUtilities.psm1

```diff
@@ -0,0 +1,310 @@
+        [pscredential] $AzsAdminCred = $null,
+
+        # The directory tenant identifier of Azure Stack Administrator.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $AzsAdminDirectoryTenantId,
+
+        # The Azure Stack ARM endpoint URI.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [Uri] $AzsAdminArmEndpoint,
+
+        # The user principal name of the account who's user data should be cleared.
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $UserPrincipalName,
+
+        # The directory tenant identifier of account who's user data should be cleared.
+        # If it is not specified, it will delete all the 
+        [Parameter(Mandatory=$false)]
+        [ValidateNotNullOrEmpty()]
+        [string] $DirectoryTenantId
+    )
+    #requires -Version 4.0
+    #requires -Module "AzureRM.Profile"
+    #requires -Module "Azs.Subscriptions.Admin"
+    #requires -RunAsAdministrator
+
+    Import-Module $PSScriptRoot\..\..\Identity\GraphAPI\GraphAPI.psm1 -Force
+
+    Write-Verbose "Login to Azure Stack Admin ARM..." -Verbose
+    $AzsAdminEnvironmentName = "AzureStackAdmin"
+    $adminArmEnv = Add-AzureRmEnvironment -Name ($AzsAdminEnvironmentName) -ARMEndpoint $AzsAdminArmEndpoint
+    Write-Verbose "Created admin ARM env as $(ConvertTo-JSON $adminArmEnv)" -Verbose
+    $loginInfo = Add-AzureRmAccount -EnvironmentName $AzsAdminEnvironmentName -Credential $AzsAdminCred -TenantId $AzsAdminDirectoryTenantId
+    Write-Verbose "Login into admin ARM with $(ConvertTo-JSON $loginInfo)" -Verbose
+
+    $adminSubscriptionId = (Get-AzureRmSubscription -Verbose | where { $_.Name -ieq "Default Provider Subscription" }).Id
+    if (-not $adminSubscriptionId)
+    {
+        throw "There is no Default Provider Subscription under the user account. Please use the Azure Stack admin account to clear user data"
+    }
+
+    if ($DirectoryTenantId)
+    {
+        $directoryTenantIdsArray = [string[]]$DirectoryTenantId
+    }
+    else 
+    {
+        Write-Verbose "Input parameter 'DirectoryTenantId' is empty. Retrieving all the registered tenant directory..." -Verbose
+        $directoryTenantIdsArray = (Get-AzsDirectoryTenant -Verbose).TenantId
+    }
+
+    Write-Host "Clearing the user data with input user principal name $UserPrincipalName and directory tenants '$DirectoryTenantIdsArray'..."
+
+    $clearUserDataResults = @() # key is directory Id, value is clear response
+    foreach ($dirId in $directoryTenantIdsArray)
+    {
+        Write-Verbose "Intializing graph env..." -Verbose
+        $graphEnvInfo = Initialize-GraphEnvironment -DirectoryTenantId $dirId -UserCredential $AadGraphCred -Environment $Environment
+        Write-Verbose "Intialized graph env with $(ConvertTo-JSON $graphEnvInfo)" -Verbose
+
+        Write-Verbose "Querying all users..." -Verbose
+        $usersResponse = Invoke-GraphApi -ApiPath "/users"
```

Comment:
How many pages does this query? It should be possible to use an ODATA filter, but might not be such a big deal without it to download all user objects

Reply:
change to the ODATA filter